eskip/eskip.go

```diff
@@ -21,6 +22,146 @@ var (
 	duplicateMethodPredicateError   = errors.New("duplicate method predicate")
 )
 
+// NewEditor creates an Editor PreProcessor, that matches routes and
+// replaces the content. For example to replace Source predicates with
+// ClientIP predicates you can use
+// --edit-route='/Source[(](.*)[)]/ClientIP($1)/', which will change
+// routes as you can see:
+//
+//        # input
+//        r0: Source("127.0.0.1/8", "10.0.0.0/8") -> inlineContent("OK") -> <shunt>
+//        # actual route
+//        edit_r0: ClientIP("127.0.0.1/8", "10.0.0.0/8") -> inlineContent("OK") -> <shunt>
+func NewEditor(reg *regexp.Regexp, repl string) *Editor {
+	return &Editor{
+		reg:  reg,
+		repl: repl,
+	}
+}
+
+type Editor struct {
+	reg  *regexp.Regexp
+	repl string
+}
+
+// NewClone creates a Clone PreProcessor, that matches routes and
+// replaces the content of the cloned routes. For example to migrate from Source to
+// ClientIP predicates you can use
+// --clone-route='/Source[(](.*)[)]/ClientIP($1)/', which will change
+// routes as you can see:
+//
+//        # input
+//        r0: Source("127.0.0.1/8", "10.0.0.0/8") -> inlineContent("OK") -> <shunt>
+//        # actual route
+//        clone_r0: ClientIP("127.0.0.1/8", "10.0.0.0/8") -> inlineContent("OK") -> <shunt>
+//        r0: Source("127.0.0.1/8", "10.0.0.0/8") -> inlineContent("OK") -> <shunt>
+func NewClone(reg *regexp.Regexp, repl string) *Clone {
+	return &Clone{
+		reg:  reg,
+		repl: repl,
+	}
+}
+
+type Clone struct {
+	reg  *regexp.Regexp
+	repl string
+}
+
+func (e *Editor) Do(routes []*Route) []*Route {
+	if e.reg == nil {
+		return routes
+	}
+
+	for i, r := range routes {
+		rr := new(Route)
+		*rr = *r
```

Comment:
just like in the Clone, we should use the canonical form rr here, as well

Reply:
fixed